contributors/design-proposals/storage/local-volume-management.md

```diff
@@ -0,0 +1,219 @@
+		- [A/D Controller Refactor](#k8s_ad_controller_refactor)
+	- [Volume Plugin Side Refactor](#volume_plugin_refactor)
+		- [Mount Device Operation](#volume_plugin_mount_refactor)
+		- [Umount Device Operation](#volume_plugin_umount_refactor)
+
+## <a name="background"></a>Background
+
+For all attachable volume plugins, there are three major steps when provisioning a volume for a `Pod`:
+
+1. Attach the volume to according node the pod is scheduled to
+2. Mount the attached device to a global mount path
+3. Bind the global mount path to a pod specified path
+
+The recycle of a volume is similar, umount/detach the path/device in a reverse order.
+
+Today, A/D controller is responsible for the attach/detach operation of volumes, kubelet is responsible for the other two steps. Before we start, let's make a quick review of the whole workflow:
+
+1. A/D controller watches all scheduled `Pods` from apiserver, and add all volumes in `Pod.Spec.Volumes` to its `DesiredStateOfWorld`. This cache maintains volumes need to be attached
+2. A/D controller periodically checks volumes in `DesiredStateOfWorld` to make sure if they are attached. If not, controller will invoke the attach operation of this volume. After attach operation finished, controller will add this volume to `Node.Status.VolumesAttached`
+3. Kubelet watches all pod scheduled to its own, and add all volumes in `Pod.Spec.Volumes` to its `DesiredStateOfWorld`
+4. Kubelet periodically checks volumes in `DesiredStateOfWorld` to make sure if they are mounted. If not, kubelet will:
+	- Wait for the attach operation finished
+	- Mount the attached device to the global mount path
+	- Bind the global mount path to a pod specified path
+
+This workflow if nicely for all cloud volumes(`AWS EBS`, `GCE PD`, etc.) which have a remote attach API, but not very suitable for local volumes(`fc`, `iscsi`, `rbd`, etc.) which need a local attach operation. So current implementation make the `Attach/Detach` method of these local volumes do nothing, and execute the real attach operations in `WaitForAttach` method. This led to the following problems:
+
+1. Device path missed in `Node.Status.VolumesAttached`.  A/D controller gets volumes' device path from the returned parameter of `Attach` method, and  updates according Node API object. Local volumes' `Attach` method does nothing, so the returned device path is empty. These absent fields may make user confused.
+
+2. Unexpected umount behavior in kubelet. Kubelet will umount the device from node if no pods use a volume, the `UnmountDeviceFunc` will use the device path stored in `Node.Status.VolumesAttached` to check if this device is still opened or not. As this device path is empty for local volumes, there may be unexpected behaviors.
+
+3. Complex and confusing code both in kubernetes side(`kubelet Volumemanager` and `A/D controller`) and volume plugin side, as local volumes have to implement `AttachableVolumePlugin` interface.
```

Comment:
This PR does not propose any solution to this complicated code, it will remain roughly the same.